File: packages/python/src/pharos_discovery/security/key_pinning.py
"""Key pinning — verify server public keys against pinned SHA-256 hashes."""

from __future__ import annotations

import hashlib
import json
import threading
from pathlib import Path

from pharos_discovery.errors import SignatureVerificationFailed


class KeyPinStore:
    """Persisted store of pinned public-key hashes keyed by server ID.

    Call :meth:`pin` to record a server's public-key SHA-256 hash, then
    :meth:`verify` with the raw public key to confirm it matches.
    Mismatches raise :class:`SignatureVerificationFailed`.
    """

    def __init__(self) -> None:
        self._pins: dict[str, str] = {}
        self._lock = threading.Lock()

    # ------------------------------------------------------------- hashing
    @staticmethod
    def hash_public_key(public_key: str) -> str:
        """Return the hex SHA-256 digest of *public_key*."""
        return hashlib.sha256(public_key.encode("utf-8")).hexdigest()

    # ------------------------------------------------------------------ API
    def pin(self, server_id: str, public_key_hash: str) -> None:
        """Pin *public_key_hash* for *server_id*."""
        with self._lock:
            self._pins[server_id] = public_key_hash

    def unpin(self, server_id: str) -> None:
        """Remove a pin (no-op if absent)."""
        with self._lock:
            self._pins.pop(server_id, None)

    def is_pinned(self, server_id: str) -> bool:
        with self._lock:
            return server_id in self._pins

    def verify(self, server_id: str, public_key: str) -> bool:
        """Verify *public_key* for *server_id*.

        Returns ``True`` if the key matches the pin. Raises
        :class:`SignatureVerificationFailed` if a pin exists but the key
        differs. Returns ``True`` if no pin exists (no pin = no constraint).
        """
        with self._lock:
            pinned = self._pins.get(server_id)
        if pinned is None:
            return True
        actual = self.hash_public_key(public_key)
        if actual != pinned:
            raise SignatureVerificationFailed(
                f"Public key hash mismatch for server '{server_id}': "
                f"expected {pinned}, got {actual}"
            )
        return True

    # ------------------------------------------------------------ persistence
    def save(self, path: str) -> None:
        with self._lock:
            data = dict(self._pins)
        Path(path).write_text(json.dumps(data, indent=2, sort_keys=True))

    @classmethod
    def load(cls, path: str) -> "KeyPinStore":
        inst = cls()
        p = Path(path)
        if p.exists():
            data = json.loads(p.read_text())
            with inst._lock:
                inst._pins = {str(k): str(v) for k, v in data.items()}
        return inst

    # ----------------------------------------------------------- dunders
    def __len__(self) -> int:
        with self._lock:
            return len(self._pins)

    def __contains__(self, server_id: object) -> bool:
        return self.is_pinned(str(server_id))

    def __repr__(self) -> str:  # pragma: no cover - trivial
        with self._lock:
            return f"KeyPinStore(pins={len(self._pins)})"

File: packages/typescript/src/security/key-pinning.ts
/**
 * Key pinning — verify server public keys against pinned SHA-256 hashes.
 *
 * Uses the Web Crypto API (`crypto.subtle.digest`) so it works in both
 * browser and Node 18+ environments without `node:` imports.
 */

import { SignatureVerificationFailed } from "../errors.js";

/** Dynamic fs loader — avoids static `node:` imports for browser compat. */
async function loadFs(): Promise<{
  writeFile(path: string, data: string, encoding: string): Promise<void>;
  readFile(path: string, encoding: string): Promise<string>;
}> {
  const mod = "node:fs/promises";
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return await (import(mod) as Promise<any>);
}

export class KeyPinStore {
  private pins = new Map<string, string>();

  /** SHA-256 of `publicKey`, returned as hex. */
  static async hashPublicKey(publicKey: string): Promise<string> {
    const enc = new TextEncoder();
    const buf = await crypto.subtle.digest("SHA-256", enc.encode(publicKey));
    return Array.from(new Uint8Array(buf))
      .map((b) => b.toString(16).padStart(2, "0"))
      .join("");
  }

  /** Pin a pre-computed hash for `serverId`. */
  pin(serverId: string, publicKeyHash: string): void {
    this.pins.set(serverId, publicKeyHash);
  }

  /** Pin the hash of `publicKey` for `serverId` (computes hash). */
  async pinKey(serverId: string, publicKey: string): Promise<string> {
    const hash = await KeyPinStore.hashPublicKey(publicKey);
    this.pins.set(serverId, hash);
    return hash;
  }

  /** Remove a pin (no-op if absent). */
  unpin(serverId: string): void {
    this.pins.delete(serverId);
  }

  /** Whether `serverId` has a pinned key. */
  isPinned(serverId: string): boolean {
    return this.pins.has(serverId);
  }

  /** Pinned hash for `serverId` (or `undefined`). */
  getPin(serverId: string): string | undefined {
    return this.pins.get(serverId);
  }

  get size(): number {
    return this.pins.size;
  }

  /**
   * Verify `publicKey` for `serverId`.
   *
   * Returns `true` if the key matches the pin, or if no pin exists (no pin =
   * no constraint).  Throws `SignatureVerificationFailed` on mismatch.
   */
  async verify(serverId: string, publicKey: string): Promise<boolean> {
    const pinned = this.pins.get(serverId);
    if (pinned === undefined) return true;
    const actual = await KeyPinStore.hashPublicKey(publicKey);
    if (actual !== pinned) {
      throw new SignatureVerificationFailed(
        `Public key hash mismatch for server '${serverId}': expected ${pinned}, got ${actual}`,
      );
    }
    return true;
  }

  /** Persist pins to `path` as JSON (Node only). */
  async save(path: string): Promise<void> {
    const data: Record<string, string> = {};
    for (const [k, v] of this.pins) data[k] = v;
    const fs = await loadFs();
    await fs.writeFile(path, JSON.stringify(data, null, 2), "utf-8");
  }

  /** Load pins from `path` (returns empty instance if missing). */
  static async load(path: string): Promise<KeyPinStore> {
    const inst = new KeyPinStore();
    try {
      const fs = await loadFs();
      const raw = await fs.readFile(path, "utf-8");
      const data = JSON.parse(raw) as Record<string, string>;
      for (const [k, v] of Object.entries(data)) inst.pins.set(k, v);
    } catch {
      // missing — return empty
    }
    return inst;
  }

  /** In-memory load from a JSON string. */
  static fromJSON(json: string): KeyPinStore {
    const inst = new KeyPinStore();
    try {
      const data = JSON.parse(json) as Record<string, string>;
      for (const [k, v] of Object.entries(data)) inst.pins.set(k, v);
    } catch {
      // ignore
    }
    return inst;
  }

  toJSON(): string {
    const data: Record<string, string> = {};
    for (const [k, v] of this.pins) data[k] = v;
    return JSON.stringify(data, null, 2);
  }
}
